File: 121.Best_Time_to_Buy_and_Sell_Stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int currMax = 0;
        int resMax = 0;
        for (int i=1; i<prices.length; i++) {
            currMax = Math.max(0, currMax + prices[i]-prices[i-1]);
            resMax = Math.max(resMax, currMax);
        }
        return resMax;
    }
    
    //brutal force
    public int maxProfit(int[] prices) {
        int size = prices.length;
        int max = 0;
        for (int i=0; i<size; i++) {
            for (int j=i+1; j<size; j++) {
                max = Math.max(max, prices[j]-prices[i]);
            }
        }
        return max;
    }
}

File: 121.BestTimetoBuyandSellStock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        res = 0
        minPrice = float('inf')
        maxProfit = 0
        for price in prices:
            minPrice = min(minPrice, price)
            maxProfit = max(maxProfit, price - minPrice)
        return maxProfit

    # can buy multiple and sell multiple
    def maxProfitAlter(self, prices: List[int]) -> int:
        res = 0
        currMax = 0
        for i in range(1, len(prices)):
            currMax = max(0, currMax + prices[i]-prices[i-1])
            res = max(res, currMax)
        return res
